exchange/accounts/balance.go

```diff
@@ -0,0 +1,169 @@
+	var curr currency.Code
+	switch v := cAny.(type) {
+	case string:
+		curr = currency.NewCode(v)
+	case currency.Code:
+		curr = v
+	default:
+		panic(fmt.Errorf("%w: '%v'", currency.ErrCurrencyNotSupported, cAny))
+	}
+	if b.Currency != currency.EMPTYCODE && b.Currency != curr {
+		panic(fmt.Errorf("%w: '%v'", errBalanceCurrencyMismatch, curr))
+	}
+	if e, ok := (*c)[curr]; !ok {
+		b.Currency = curr
+		(*c)[curr] = b
+	} else {
+		(*c)[curr] = e.Add(b)
+	}
+}
+
+// Balance returns a snapshot copy of the Balance
+func (b *balance) Balance() Balance {
+	b.m.RLock()
+	defer b.m.RUnlock()
+	return b.internal
+}
+
+// Add returns a new Balance adding together a and b
+// UpdatedAt is the later of the two Balances
+func (b *Balance) Add(a Balance) Balance { //nolint:gocritic // hugeparam not relevant; We'd need to copy it in map iterations anyway
+	var u time.Time
+	if a.UpdatedAt.After(b.UpdatedAt) {
+		u = a.UpdatedAt
+	} else {
+		u = b.UpdatedAt
+	}
+	return Balance{
+		Total:                  b.Total + a.Total,
+		Hold:                   b.Hold + a.Hold,
+		Free:                   b.Free + a.Free,
+		AvailableWithoutBorrow: b.AvailableWithoutBorrow + a.AvailableWithoutBorrow,
+		Borrowed:               b.Borrowed + a.Borrowed,
+		UpdatedAt:              u,
+	}
+}
+
+func (c *currencyBalances) Public() CurrencyBalances {
+	n := make(CurrencyBalances, len(*c))
+	for curr, bal := range *c {
+		n[curr] = bal.Balance()
+	}
+	return n
+}
+
+// update checks that an incoming change has a valid change, and returns if the balances were changed
+// If change does not have a Currency set, the existing Currency is preserved
+func (b *balance) update(change Balance) (bool, error) { //nolint:gocritic // hugeparam not relevant; We'd need to copy it later anyway
+	if err := common.NilGuard(b); err != nil {
+		return false, err
+	}
+	if change.UpdatedAt.IsZero() {
+		return false, errUpdatedAtIsZero
+	}
+	b.m.Lock()
+	defer b.m.Unlock()
+	if b.internal.Currency != currency.EMPTYCODE {
+		switch change.Currency {
+		case b.internal.Currency:
+			// All good
+		case currency.EMPTYCODE:
+			change.Currency = b.internal.Currency
+		default:
+			return false, errBalanceCurrencyMismatch
+		}
+	}
+	if !b.internal.UpdatedAt.Before(change.UpdatedAt) {
+		return false, errOutOfSequence
+	}
+	b.internal.UpdatedAt = change.UpdatedAt // Set just the time, and then can compare easily
+	if b.internal == change {
+		return false, nil
+	}
+	b.internal = change
+	return true, nil
+}
+
+// balance rutens a balance for a currency
+func (c currencyBalances) balance(curr currency.Code) *balance {
+	if _, ok := c[curr]; !ok {
+		c[curr] = &balance{internal: Balance{Currency: curr}}
+	}
+	return c[curr]
+}
+
+func (c CurrencyBalances) clone() CurrencyBalances {
+	n := make(CurrencyBalances, len(c))
+	for curr, bal := range c {
+		n[curr] = bal
+	}
+	return n
+}
```

Comment:
RM and rm test as its only being used in the other `(l SubAccounts) clone()` func which isn't being used outside of a test.

Reply:
Fixed https://github.com/gbjk/gocryptotrader/commit/b0df20ba6